runtime/compiler/control/CompilationThread.cpp

```diff
@@ -8201,7 +8269,13 @@ TR::CompilationInfoPerThreadBase::wrappedCompile(J9PortLibrary *portLib, void *
                   (!TR::Compiler->target.cpu.isPower() && // Temporary change until we figure out the AOT bug on PPC
                    !TR::Options::getAOTCmdLineOptions()->getOption(TR_DisableAotAtCheapWarm))) &&
                   p->_optimizationPlan->isOptLevelDowngraded() &&
-                  p->_optimizationPlan->getOptLevel() == cold) // Is this test really needed?
+                  p->_optimizationPlan->getOptLevel() == cold // Is this test really needed?
+#if defined(J9VM_OPT_JITSERVER)
+                  // Do not reupgrade a compilation that was downgraded due to low memory
+                  && (TR::Options::getCmdLineOptions()->getOption(TR_DisableJITServerBufferedExpensiveCompilations) ||
+                      !that->_methodBeingCompiled->shouldUpgradeOutOfProcessCompilation())
```

Comment:
For AOT compilations we start at cold, so `downgradeLocalCompIfLowPhysicalMemory()` will not do any downgrade, which means that in this code we will upgrade back

Reply:
Hm, I ran my tests with `dontDowngradeToCold`, I suppose that makes it start AOT compiles at warm, so I didn't see this issue.
Does this mean that if an AOT compilation starts at cold, it will still be marked as downgraded and upgraded to warm here? Because this if statement also checks that the compilation is downgraded, so I thought if we start at cold, then we continue at cold, and this upgrade is not triggered.

Reply:
The sequence of actions is the following:
- JIT detects start-up phase and downgrade the opt level at cold
- JIT decides whether or not to use AOT. Heuristic says that cold compilations should be performed with AOT (the logic behind this is that, we downgraded to cold in order to improve start-up time, so why not go the extra mile and generate AOT to further improve start-up time)
- Later-on the JIT decides to re-upgrade the AOT compilation at cheap-warm to generate more optimized code.

If you run with `dontDowngradeToCold` there should be no AOT, unless `-Xaot:forceaot` is used

Reply:
I see. Indeed, I've been running with `forceAOT`.
So to avoid upgrades, do we want to measure free memory here as well?

Reply:
What we need is for `_methodBeingCompiled->shouldUpgradeOutOfProcessCompilation()` to answer `true` even for AOT compilations. This means that `downgradeLocalCompilationIfLowPhysicalMemory()` needs to process AOT compilations specially. The test should become:
```
   if (!TR::Options::getCmdLineOptions()->getOption(TR_DisableJITServerBufferedExpensiveCompilations) &&
       TR::Options::getCmdLineOptions()->allowRecompilation() &&
       !TR::CompilationInfo::isCompiled(method) &&  // do not downgrade recompilations
       (entry->_optimizationPlan->getOptLevel() >= warm || 
        // AOT compilations that are upgraded back to cheap-warm should be considered expensive
         (entry->_useAOTCompilation && !TR::Options::getAOTCmdLineOptions()->getOption(TR_DisableAotAtCheapWarm))
      )
```